src/repo/kv.rs

```diff
@@ -0,0 +1,387 @@
+use super::{Column, DataStore};
+use crate::error::Error;
+use crate::repo::{PinKind, PinMode, PinStore, References};
+use async_trait::async_trait;
+use cid::{self, Cid};
+use futures::stream::{StreamExt, TryStreamExt};
+use sled::{self, Config as DbConfig, Db, Mode as DbMode};
+use std::convert::Into;
+use std::path::PathBuf;
+use std::str::{self, FromStr};
+use std::sync::Mutex;
+use tracing_futures::Instrument;
+
+#[derive(Debug)]
+pub struct KvDataStore {
+    path: PathBuf,
+    db: Mutex<Option<Db>>,
+}
+
+impl KvDataStore {
+    fn _put(&self, key: &str, value: &str) -> Result<(), Error> {
+        let db = self.get_db();
+
+        let _ = db.insert(key, value)?;
+
+        Ok(())
+    }
+
+    fn _remove(&self, key: &str) -> Result<(), Error> {
+        let db = self.get_db();
+
+        match db.remove(key) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    fn _apply_batch(&self, batch: sled::Batch) -> Result<(), Error> {
+        let db = self.get_db();
+
+        Ok(db.apply_batch(batch)?)
+    }
+
+    fn get_db(&self) -> Db {
+        self.db.lock().unwrap().as_ref().unwrap().clone()
+    }
+}
+
+#[async_trait]
+impl DataStore for KvDataStore {
+    fn new(root: PathBuf) -> KvDataStore {
+        KvDataStore {
+            path: root,
+            db: Default::default(),
+        }
+    }
+
+    async fn init(&self) -> Result<(), Error> {
+        let config = DbConfig::new();
+
+        let db = config
+            .mode(DbMode::HighThroughput)
+            .path(self.path.as_path())
+            .open()?;
```

Comment:
since there is a `DataStore::open`, maybe that one should perform this action instead?

Reply:
I think the `open` needs to be thrown away and I don't think they are even called, hopefully even along with `init` and go full RAII with these components. They are mostly useless after being new'd without some interior mutability going on.